Objects/regions.c

```diff
@@ -1337,12 +1429,146 @@ int _Py_is_bridge_object(PyObject *op) {
     return ((Py_region_ptr_t)((regionmetadata*)region)->bridge == (Py_region_ptr_t)op);
 }
 
-static void regionmetadata_close(regionmetadata* data) {
-    data->is_open = 0;
-}
+static int try_close(PyRegionObject *root_bridge) {
+    addtoregionvisitinfo info = {
+        .pending = stack_new(),
+        .new_sub_regions = stack_new(),
+        // `src` is reassigned each iteration
+        .src = NULL,
+    };
+    if (!info.pending || !info.new_sub_regions) {
+        goto fail;
+    }
 
-static regionmetadata* PyRegion_get_metadata(PyRegionObject* obj) {
-    return obj->metadata;
+    if (stack_push(info.new_sub_regions, _PyObject_CAST(root_bridge))) {
+        PyErr_NoMemory();
+        goto fail;
+    }
+
+    // The root region can have to have two local references, one from the
+    // owning reference and one from the `self` argument
+    Py_ssize_t root_region_lrc_limit;
+    regionmetadata *root_data = Py_REGION_DATA(root_bridge);
+    if (regionmetadata_has_parent(root_data) || root_data->cown) {
+        root_region_lrc_limit = 1;
+    } else {
+        root_region_lrc_limit = 2;
+    }
+
+    while (!stack_empty(info.new_sub_regions)) {
+        PyObject *bridge = stack_pop(info.new_sub_regions);
+        assert(Py_is_bridge_object(bridge));
+        regionmetadata* old_data = Py_REGION_DATA(bridge);
+
+        // One from the owning reference
+        Py_ssize_t rc_limit = 1;
+        Py_ssize_t lrc_limit = 0;
+
+        // The root bridge has different limits since it's currently used
+        // as an argument for this method.
+        if (bridge == _PyObject_CAST(root_bridge)) {
+            rc_limit += 1;
+            lrc_limit = root_region_lrc_limit;
+        }
+
+        // The *LRC* and *is_open* status is currently not updated when references
+        // to the bridge are created. This means that the bridge might have multiple
+        // unknown references.
+        //
+        // Using the object RC is an over approximation, since internal cycles from
+        // objects to the bridge object will also increase the RC thereby tricking
+        // this check into opening it again.
+        if (Py_REFCNT(bridge) > rc_limit) {
+            regionmetadata_open(Py_REGION_DATA(bridge));
+        }
+
+        // If it's closed there is nothing we need to do.
+        if (!regionmetadata_is_open(old_data)) {
+            continue;
+        }
+
+        // Create the new `regionmetadata*`
+        regionmetadata* new_data = (regionmetadata*)calloc(1, sizeof(regionmetadata));
+        if (!new_data) {
+            PyErr_NoMemory();
+            goto fail;
+        }
+        _Py_CAST(PyRegionObject *, bridge)->metadata = new_data;
+        regionmetadata_inc_rc(new_data);
```

Comment:
Is the inc here because of assigning the metadata?  Does this mean you also need a decref on the old data?  Or is that being transferred to the merge?  I think some comments on the RC usage here, or macros/functions to combine would be helpful for readability here.

Reply:
Good catch, yeah pretty sure it needs to dec rc here

Reply:
And I also added comments